alf/utils/math_ops.py

```diff
@@ -82,6 +83,114 @@ def swish(x):
+
+
+@alf.configurable
+def softupper(x, high, hinge_softness=1.):
+    """Softly upper bound ``x`` by ``high``, namely,
+    ``softupper(x, high) = -softplus(high - x) + high``.
+
+    Args:
+        x (Tensor): input
+        high (float|Tensor): the upper bound
+        hinge_softness (float): this positive parameter changes the transition
+            slope. A higher softness results in a smoother transition from
+            identity to ``high``. Default to 1.
+
+    Returns:
+        Tensor
+    """
+    assert hinge_softness > 0
+    return -nn.functional.softplus(high - x, beta=1. / hinge_softness) + high
+
+
+@alf.configurable
+def softclip_tf(x, low, high, hinge_softness=1.):
+    """Softly bound ``x`` in between ``[low, high]``, namely,
+
+    .. code-block:: python
+
+        clipped = softupper(softlower(x, low), high)
+        softclip(x) = (clipped - high) / (high - softupper(low, high)) * (high - low) + high
+
+    The second scaling step is because we will have
+    ``softupper(low, high) < low`` due to distortion of softplus, so we need to
+    shrink the interval slightly by ``(high - low) / (high - softupper(low, high))``
+    to preserve the lower bound. Due to this rescaling, the bijector can be mildly
+    asymmetric.
+
+    Args:
+        x (Tensor): input
+        low (float|Tensor): the lower bound
+        high (float|Tensor): the upper bound
+        hinge_softness (float): this positive parameter changes the transition
+            slope. A higher softness results in a smoother transition from
+            ``low`` to ``high``. Default to 1.
+    """
+    if not isinstance(low, torch.Tensor):
+        low = torch.tensor(low)
+    assert torch.all(high > low), "Invalid clipping range"
+
+    softupper_high_at_low = softupper(low, high, hinge_softness)
+    clipped = softupper(
+        softlower(x, low, hinge_softness), high, hinge_softness)
+    return ((clipped - high) / (high - softupper_high_at_low) * (high - low) +
+            high)
+
+
+@alf.configurable
+def softclip(x, low, high, hinge_softness=1., threshold=10.):
+    r"""Softly bound ``x`` in between ``[low, high]``. Unlike ``softclip_tf``,
+    this transform is symmetric regarding the lower and upper bound when
+    squashing. The softclip function can be defined in several forms:
+
+    .. math::
+
+        \begin{array}{lll}
+            &\ln(\frac{e^{l-x}+1}{e^{x-h}+1}) + x & (1)\\
+            =&\ln(\frac{e^{x-l}+1}{e^{x-h}+1}) + l & (2)\\
+            =&\ln(\frac{e^{l-x}+1}{e^{h-x}+1}) + h & (3)\\
+        \end{array}
+
+    Args:
+        x (Tensor): input
+        low (float|Tensor): the lower bound
+        high (float|Tensor): the upper bound
+        hinge_softness (float): this positive parameter changes the transition
+            slope. A higher softness results in a smoother transition from
+            ``low`` to ``high``. Default to 1.
+        threshold (float): For numerical stability, the function becomes (2) when
```

Comment:
why is this parameter needed? It seems to me that threshold=0. is always a better choice because (1) and (2) is guaranteed to be exactly same at low and (1) and (3) is guaranteed to be exactly same at high even considering the finite precision calculation.

Reply:
Removed